jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -1771,14 +1793,25 @@ public boolean equalsIgnoreCase(String string) {
 
 			if (charAtO1Last != charAtO2Last &&
 					toUpperCase(charAtO1Last) != toUpperCase(charAtO2Last) &&
-					((charAtO1Last <= 255 && charAtO2Last <= 255) || Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last))) {
+					((charAtO1Last <= 255 && charAtO2Last <= 255) || Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last)) &&
+					(!Character.isLowSurrogate(charAtO1Last) || !Character.isLowSurrogate(charAtO2Last))) {
 				return false;
 			}
 
 			while (o1 < end - 1) {
 				char charAtO1 = s1.charAtInternal(o1++, s1Value);
 				char charAtO2 = s2.charAtInternal(o2++, s2Value);
 
+				/*[IF Java16]*/
+				if (Character.isHighSurrogate(charAtO1) && Character.isHighSurrogate(charAtO2)) {
+					int codepointAtO1 = Character.toCodePoint(charAtO1, s1.charAtInternal(o1++, s1Value));
+					int codepointAtO2 = Character.toCodePoint(charAtO2, s2.charAtInternal(o2++, s2Value));
+					if ((codepointAtO1 != codepointAtO2) && (Character.toUpperCase(codepointAtO1) != Character.toUpperCase(codepointAtO2))) {
```

Comment:
Shouldn't this be doing Character.toUpperCase(Character.toLowerCase(codepoint)) similarly to compareValue()?

Reply:
Yes, will fix here and in `regionMatches`.